target/linux/ramips/patches-5.10/415-add-support-for-eon-en25qh256a.patch

```diff
@@ -0,0 +1,42 @@
+From 6abef37d16d0c570ef5a149e63762fba2a30804b Mon Sep 17 00:00:00 2001
+From: "Leon M. George" <leon@georgemail.eu>
+Date: Wed, 30 Mar 2022 16:16:56 +0200
+Subject: [PATCH] mtd: spi-nor: support eon en25qh256a variant
+
+The en25qh256a requires 4-byte_addr_mode to operate.
+Otherwise, it appears to be identical to the en25qh256.
+
+Signed-off-by: Leon M. George <leon@georgemail.eu>
+---
+ drivers/mtd/spi-nor/eon.c | 11 +++++++++++
+ 1 file changed, 11 insertions(+)
+
+diff --git a/drivers/mtd/spi-nor/eon.c b/drivers/mtd/spi-nor/eon.c
+index 4f3ee6331f37..89be24f42225 100644
+--- a/drivers/mtd/spi-nor/eon.c
++++ b/drivers/mtd/spi-nor/eon.c
+@@ -8,6 +8,15 @@
+ 
+ #include "core.h"
+ 
++static void en25qh256a_default_init(struct spi_nor *nor)
++{
++	nor->params->set_4byte_addr_mode = spi_nor_set_4byte_addr_mode;
++}
++
++static const struct spi_nor_fixups en25qh256a_fixups = {
++	.default_init = en25qh256a_default_init,
++};
++
+ static const struct flash_info eon_parts[] = {
+ 	/* EON -- en25xxx */
+ 	{ "en25f32",    INFO(0x1c3116, 0, 64 * 1024,   64, SECT_4K) },
+@@ -24,6 +33,8 @@ static const struct flash_info eon_parts
+ 	{ "en25qh64",   INFO(0x1c7017, 0, 64 * 1024,  128,
+ 			     SECT_4K | SPI_NOR_DUAL_READ) },
+ 	{ "en25qh128",  INFO(0x1c7018, 0, 64 * 1024,  256, 0) },
++	{ "en25qh256a",  INFO(0x1c7019, 0x1c70, 64 * 1024,  512, 0)
++		.fixups = &en25qh256a_fixups },
+ 	{ "en25qh256",  INFO(0x1c7019, 0, 64 * 1024,  512, 0) },
```

Comment:
glad you discovered the boot problem, you are on the right track, however...

A duplicate ID will never be accepted, neither by openwrt or upstream linux

the whole point of making a "fixup" function is to handle 2 cases with the same JEDEC ID in the same chip INFO definition.

I have seen a similar situation with another pair of flash chips, and this is how they handled it

https://elixir.bootlin.com/linux/v5.10.110/C/ident/mx25l25635_post_bfpt_fixups

for reference, I did find the datasheets for both variants, and I can confirm that one supports 4-byte codes and the other doesnt. 

https://esmt.com.tw/upload/pdf/ESMT/datasheets/EN25QH256_Ver.G.pdf
https://esmt.com.tw/upload/pdf/ESMT/datasheets/EN25QH256A(2R).pdf

by the way, setting 4-byte only mode for the fixup is not exactly appropriate either, probably the fixup function just needs to decide whether or not to add the flag `SNOR_F_4B_OPCODES` (4-byte instruction set support)

Reply:
...seems that elixir is broken at the moment...

anyway here is the commit for the example im talking about

https://github.com/torvalds/linux/commit/2bffa65da43e399079dad5947c6aa9ab3cfa4ad4

Reply:
I read the datasheets a little, see if it works if the fixup function is like this...

taking this directly from the commit, and using a bit in DWORD 1 of the SFDP info instead of a bit in DWORD 5

remember that, there should be only 1 definition with the same ID, including this function as a fixup

```
static int
en25qh256_post_bfpt_fixups(struct spi_nor *nor,
			    const struct sfdp_parameter_header *bfpt_header,
			    const struct sfdp_bfpt *bfpt,
			    struct spi_nor_flash_parameter *params)
{
	/*
	 * long comment to explain
	 */
	if (bfpt->dwords[BFPT_DWORD(1)] & BFPT_DWORD1_FAST_READ_1_1_4)
		nor->flags |= SNOR_F_4B_OPCODES;

	return 0;
}

static struct spi_nor_fixups en25qh256_fixups = {
	.post_bfpt = en25qh256_post_bfpt_fixups,
};
```

Reply:
> A duplicate ID will never be accepted, neither by openwrt or upstream linux

That sound entirely reasonable to do. Thank you for pointing out how it should be dealt with!

> I have seen a similar situation with another pair of flash chips, and this is how they handled it

I will have a look in a moment.

> remember that, there should be only 1 definition with the same ID, including this function as a fixup

Can do - Will do right after fetching the boot log.

Reply:
elixir search engine is broke, look at the commit instead 👍🏼

Reply:
Looking into this, I'm starting to think that the way using `default_init` is the way to go.

The way to detect the 4-byte-address-mode - if I understand correctly - is to read `bfpt->dwords[BFPT_DWORD(1)] & BFPT_DWORD1_ADDRESS_BYTES_MASK`.

That is already done in `spi_nor_parse_bfpt` , where `nor->addr_width` is set accordingly:
```
  case BFPT_DWORD1_ADDRESS_BYTES_3_ONLY:
  case BFPT_DWORD1_ADDRESS_BYTES_3_OR_4:
    nor->addr_width = 3;
    break;
  case BFPT_DWORD1_ADDRESS_BYTES_4_ONLY:
    nor->addr_width = 4;
```

`addr_width == 4` should in turn be checked by `spi_nor_init` where `params->set_4byte_addr_mode` is called (`!(nor->flags & SNOR_F_4B_OPCODES)` is also required).
But.. by default, it's set to `spansion_set_4byte_addr_mode` - which needs to be changed to `spi_nor_set_4byte_addr_mode` (which is what the patch already does).

I can image everything just working on plain `en25qh256` (without `a`) if it doesn't set `BFPT_DWORD1_ADDRESS_BYTES_4_ONLY` but somebody would need to test that (I don't have access to that chip).

The `post_bfpt` fixup didn't get called. I've pushed an updated version of the patch to reflect the current state.
still WIP, then

Reply:
> The way to detect the 4-byte-address-mode - if I understand correctly - is to read `bfpt->dwords[BFPT_DWORD(1)] & BFPT_DWORD1_ADDRESS_BYTES_MASK`.

unfortunately no, it's never that easy 🙃 

those bits only tell you whether the chip is capable of 4-byte addressing at all, and any chip greater than 16 MB has to be, however it doesnt tell anyone HOW it uses 4-byte addressing...

the EN25QH256 and EN25QH256A variants both have the ability to read 4-byte addresses, but they do so in different ways

for example, there are some chips that allow multiple ways to do so
(from MX25L25635F datasheet)
```
The MX25L25635F provides three different methods to access the whole 256Mb density:

 1. Command entry 4-byte address mode: Issue Enter 4-Byte mode command to set up the 4BYTE bit in
Configuration Register bit. After 4BYTE bit has been set, the number of address cycle become 32-bit.

 2. Extended Address Register (EAR): configure the memory device into two 128Mb segments to select which one
is active through the EAR bit “0”.

 3. 4-byte Address Command Set: When issuing 4-byte address command set, 4-byte address (A31-A0) is
requested after the instruction code. Please note that it is not necessary to issue EN4B command before issuing
any of 4-byte command set.
```

that last method "4-byte address command set" is what the flag `SNOR_F_4B_OPCODES` is for.

The EN25QH256 uses the first method (linux default), and the A variant uses both the first and last method (completely separate read and write instructions for the different address lengths, and a bit to change address length for the normal read and write instructions)

another thing to keep in mind, `addr_width` is a dynamic variable. Whenever a read or write is requested, that value will flip-flop between 3 and 4 depending on the offset or length of the read or write.

another problem you might be facing is whether u-boot is setting 4-byte address mode at boot, and then linux tries to read a 3-byte address....but I imagine the driver should handle that already somehow...

for debug, a useful thing would be for the fixup function to print what address mode it is in before and after

yet ANOTHER thing to consider.....

if you did a sysupgrade without powering off the router completely, it may be possible that it never reset to 3-byte addressing mode on the next boot, if there is a problem with the reset pin on the flash chip (sometimes it is not even connected)

Reply:
Again, thank you very much!

> if you did a sysupgrade without powering off the router completely, it may be possible that it never reset to 3-byte addressing mode on the next boot, if there is a problem with the reset pin on the flash chip (sometimes it is not even connected)
In the current state, the router always freezes on plain `reboot`.
The cudy firmware (which also enables 4 byte address commands) doesn't have this problems.

I should take a closer look at those datasheets now to find a decent method for distinguishing the variants and for accessing the flash.

Reply:
I've finally managed to find some more time and share my findings.

First: Yes, u-boot doesn't load if the flash was in 4-byte-mode on reset.
The cudy firmware uses the reset-hack. Likely, the reset pin isn't connected - as you say.

Here is an overview of features relevant to accessing the upper memory bank (>8m) according to the data sheets:

```
        EN25QH256 | EN25QH256A
 4ba        x     |      x      (modifies the behaviour of existing commands / causes problems on reset)
 HBL        x     |      -      (high bank latch controls whether access is below or above 8m)
READ4A      -     |      x      (op that supports 4ba natively)
4bait       -     |      x      (chip announces instructions via sfdp)
```

Notably, the HBL (opcode 0x67) appears to be different from the EAR that you've mentioned and would likely need to implemented seperately (it affects read, erase, and programm commands).

By default, `set_4byte_addr_mode` is `spansion_set_4byte_addr_mode` (`spi_nor_info_init_params`).
Setting `set_4byte_addr_mode` to `spi_nor_set_4byte_addr_mode` should be correct for both variants (but not necessarily desirable).
(Neither datasheet indicates that `spansion_set_4byte_addr_mode` would work).
Support for the EN25QH256 (no A) is merged into Mainline Linux, should be able to access the upper 8m, and, assuming it does, does so without 4ba.
`set_4byte_addr_mode` is called by `spi_nor_init` when `nor->addr_width == 4 && !(nor->flags & SNOR_F_4B_OPCODES)`.
Assuming the EN25QH256 isn't broken in Linux mainline, `SNOR_F_4B_OPCODES` must be set by `spi_nor_parse_4bait` for the EN25QH256.

Adding the flags `SPI_NOR_DUAL_READ | SPI_NOR_QUAD_READ` brings the device up and running.
It should work on both variants (according to the datasheet - naturally, it needs testing).
That also avoids the 'broken reset' issue (indeed, the reset pin isn't connected for the Cudy M1800).
When I add `SNOR_F_4B_OPCODES`, there's a crash:
<details><summary>crash log</summary>

```
[   10.431297] CPU 0 Unable to handle kernel paging request at virtual address ffffde98, epc == 82495b08, ra == 828b2fe4
[   10.431310] mt7915e: probe of 0000:02:00.0 failed with error -22
[   10.448056] Oops[#1]:
[   10.450379] CPU: 0 PID: 10 Comm: ksoftirqd/0 Tainted: G        W         5.10.107 #0
[   10.458135] $ 0   : 00000000 00000000 82495a00 828b5df8
[   10.463425] $ 4   : 826d2000 000d7010 00000000 00000001
[   10.468702] $ 8   : 8145fc80 00000000 02df1b05 00000354
[   10.473979] $12   : 81006b80 02df1b05 00000000 00000001
[   10.479256] $16   : 02df1b05 00000000 826d2000 808b67a0
[   10.484532] $20   : 807fa40c 00000000 3fffffff 807ec040
[   10.489810] $24   : 00000000 82760178
[   10.495087] $28   : 8145e000 8145fdf8 808b02e8 828b2fe4
[   10.500365] Hi    : 0000006d
[   10.503280] Lo    : 999999b5
[   10.504906] usbcore: registered new interface driver qmi_wwan
[   10.506218] epc   : 82495b08 0x82495b08
[   10.515768] ra    : 828b2fe4 mt7915_dual_hif_set_irq_mask+0x128/0x448 [mt7915e]
[   10.515862] usbcore: registered new interface driver sierra_net
[   10.523057] Status: 11008403 KERNEL EXL IE
[   10.523083] Cause : 40800008 (ExcCode 02)
[   10.523088] BadVA : ffffde98
[   10.523094] PrId  : 0001992f (MIPS 1004Kc)
[   10.523098] Modules linked in: sierra_net(+) qmi_wwan nft_reject_ipv6 nft_reject_ipv4 nft_reject_inet nft_reject nft_redir nft_quota nft_objref nft_numgen nft_nat nft_masq nft_log nft_limit nft_hash nft_flow_offload nft_ct nft_counter nft_chain_nat nf_tables nf_nat nf_flow_table nf_conntrack mt7915e mt7603e mt76_connac_lib mt76 mac80211 ftdi_sio cp210x cfg80211 cdc_ncm cdc_ether usbserial usbnet usbhid nfnetlink nf_reject_ipv6 nf_reject_ipv4 nf_defrag_ipv6 nf_defrag_ipv4 libcrc32c hwmon hid_generic compat cdc_wdm hid evdev input_core l2tp_eth l2tp_netlink l2tp_core udp_tunnel ip6_udp_tunnel veth tun vfat fat nls_utf8 nls_iso8859_1 nls_cp437 sha256_generic libsha256 seqiv jitterentropy_rng drbg hmac cmac usb_storage leds_gpio uhci_hcd ohci_platform ohci_hcd fsl_mph_dr_of ehci_platform ehci_fsl sd_mod scsi_mod ehci_hcd gpio_button_hotplug usbcore nls_base usb_common mii crc32c_generic
[   10.591066] batman_adv: B.A.T.M.A.N. advanced 2022.0-openwrt-2 (compatibility version 15) loaded
[   10.622508] Process ksoftirqd/0 (pid: 10, threadinfo=15eebd9c, task=2eea967f, tls=00000000)
[   10.622516] Stack : 810072a0 00000000
[   10.635245] usbcore: registered new interface driver cdc_mbim
[   10.639601]  80866c60 8009a064 807fa40c 826d4154 00000000 81000280
[   10.639634]         808b67a0 80034570 3fffffff 00000001
[   10.652726] usbcore: registered new interface driver option
[   10.655211]  807ec05c
[   10.660553] usbserial: USB Serial support registered for GSM modem (1-port)
[   10.665947]  807ec064
[   10.671551] usbcore: registered new interface driver qcserial
[   10.675139]  00000002
[   10.677562] usbserial: USB Serial support registered for Qualcomm USB modem
[   10.683108]  807ec058
[   10.687113] kmodloader: done loading kernel modules from /etc/modules.d/*
[   10.692281]
[   10.692286]         00000040 00000006 00000007 00000100 807fa40c 8068f5c4 809c0000 800607ac
[   10.711143]         81006b40 807fcf40 00000000 81006b40 808b0288 808b67a0 00000009 ffff8ede
[   10.719514]         80800000 80726010 04208040 8072610c 81420600 00000001 807ffc48 00000002
[   10.727885]         ...
[   10.730380] Call Trace:
[   10.730386]
[   10.734325] [<8009a064>] rcu_core+0xa4/0x638
[   10.738597] [<80034570>] tasklet_action_common.constprop.0+0x98/0x17c
[   10.745033] [<8068f5c4>] __do_softirq+0x164/0x368
[   10.749750] [<800607ac>] pick_next_task_idle+0x38/0x4c
[   10.754880] [<80034814>] run_ksoftirqd+0x5c/0x70
[   10.759484] [<8068b5c4>] schedule_preempt_disabled+0x10/0x1c
[   10.765153] [<8068b138>] schedule+0x54/0xf8
[   10.769327] [<80056b38>] smpboot_thread_fn+0x100/0x1d4
[   10.774454] [<80050978>] __kthread_parkme+0x40/0x88
[   10.779317] [<8068b5c4>] schedule_preempt_disabled+0x10/0x1c
[   10.784958] [<80056a38>] smpboot_thread_fn+0x0/0x1d4
[   10.789909] [<80050be0>] kthread+0x13c/0x144
[   10.794165] [<80050aa4>] kthread+0x0/0x144
[   10.798251] [<80050aa4>] kthread+0x0/0x144
[   10.802337] [<80050aa4>] kthread+0x0/0x144
[   10.806425] [<80002ff8>] ret_from_kernel_thread+0x14/0x1c
[   10.811816]
[   10.813300] Code: 00000000  82690000  82495a00 <8009de98> 00000000  00000000  00000000  00000000  00000000
[   10.823066]
[   10.824886] ---[ end trace 72df6ea1d4889396 ]---
[   10.829573] Kernel panic - not syncing: Fatal exception in interrupt
[   10.835945] ------------[ cut here ]------------
[   10.840596] WARNING: CPU: 0 PID: 10 at kernel/smp.c:633 smp_call_function_many_cond+0x438/0x454
[   10.849255] Modules linked in: qcserial option nf_flow_table_ipv6 nf_flow_table_ipv4 nf_flow_table_inet cdc_mbim batman_adv ath10k_pci ath10k_core ath usb_wwan sierra_net qmi_wwan nft_reject_ipv6 nft_reject_ipv4 nft_reject_inet nft_reject nft_redir nft_quota nft_objref nft_numgen nft_nat nft_masq nft_log nft_limit nft_hash nft_flow_offload nft_ct nft_counter nft_chain_nat nf_tables nf_nat nf_flow_table nf_conntrack mt7915e mt7603e mt76_connac_lib mt76 mac80211 ftdi_sio cp210x cfg80211 cdc_ncm cdc_ether usbserial usbnet usbhid nfnetlink nf_reject_ipv6 nf_reject_ipv4 nf_defrag_ipv6 nf_defrag_ipv4 libcrc32c hwmon hid_generic compat cdc_wdm hid evdev input_core l2tp_eth l2tp_netlink l2tp_core udp_tunnel ip6_udp_tunnel veth tun vfat fat nls_utf8 nls_iso8859_1 nls_cp437 sha256_generic libsha256 seqiv jitterentropy_rng drbg hmac cmac usb_storage leds_gpio uhci_hcd ohci_platform ohci_hcd fsl_mph_dr_of ehci_platform ehci_fsl sd_mod scsi_mod ehci_hcd gpio_button_hotplug usbcore nls_base usb_common
[   10.849891]  mii crc32c_generic
[   10.940092] CPU: 0 PID: 10 Comm: ksoftirqd/0 Tainted: G      D W         5.10.107 #0
[   10.947803] Stack : 809e0000 00000000 808b02e8 80083470 80870000 80765b30 00000000 00000000
[   10.956172]         8145fb44 809c0000 80734498 8143b8a8 807fce27 00000001 8145fae8 721d0047
[   10.964541]         00000000 00000000 80734498 8145f988 ffffefff 00000000 ffffffea 00000000
[   10.972909]         8145f994 00000185 808027f8 ffffffff 00000000 00000000 00000000 80730000
[   10.981276]         00000009 8143b4e0 00000000 00000000 00000018 803f53b8 00000000 809c0000
[   10.989642]         ...
[   10.992090] Call Trace:
[   10.994551] [<800080e0>] show_stack+0x30/0x100
[   10.998991] [<80372b7c>] dump_stack+0x9c/0xcc
[   11.003359] [<8002ff28>] __warn+0xc0/0x12c
[   11.007444] [<8002fff0>] warn_slowpath_fmt+0x5c/0xac
[   11.012400] [<800c127c>] smp_call_function_many_cond+0x438/0x454
[   11.018392] [<800c12dc>] smp_call_function+0x24/0x30
[   11.023344] [<8002f9ac>] panic+0x118/0x308
[   11.027431] [<800083a0>] die+0x10c/0x118
[   11.031361] [<80018d6c>] do_page_fault+0x4ac/0x4bc
[   11.036143] [<8001dfe8>] tlb_do_page_fault_0+0x118/0x120
[   11.041486] [<828b2fe4>] mt7915_dual_hif_set_irq_mask+0x128/0x448 [mt7915e]
[   11.048444] [<80034570>] tasklet_action_common.constprop.0+0x98/0x17c
[   11.054878] [<8068f5c4>] __do_softirq+0x164/0x368
[   11.059571] [<80034814>] run_ksoftirqd+0x5c/0x70
[   11.064182] [<80056b38>] smpboot_thread_fn+0x100/0x1d4
[   11.069312] [<80050be0>] kthread+0x13c/0x144
[   11.073572] [<80002ff8>] ret_from_kernel_thread+0x14/0x1c
[   11.078955]
[   11.080442] ---[ end trace 72df6ea1d4889397 ]---
```

<details>

Reply:
If the patch doesn't work on the no-A-variant, it should be possible to distinguish both variants using the `SNOR_HWCAPS_READ_1_1_4` bit (0x32, bit 7)
It's enabled on the A-variant and disabled on the other.
Another difference is the "Sector Type 2 Size" (0x4e) and "Opcode" (0x4f) - they read 0x00 and 0xFF on the no-A-variant.

Reply:
I've decided to implement the detection method for the older kernel.
It involves exporting `spi_nor_default_setup`.

```
static int en25qh256_nor_setup(struct spi_nor *nor,
          const struct spi_nor_hwcaps *hwcaps)
{
  if (hwcaps->mask & SNOR_HWCAPS_READ_1_1_4)
    nor->flags |= SNOR_F_4B_OPCODES;
  
  return spi_nor_default_setup(nor, hwcaps);
}         
          
static void en25qh256_post_sfdp_fixups(struct spi_nor *nor)
{
  nor->params->setup = en25qh256_nor_setup;
}                     
                      
static const struct spi_nor_fixups en25qh256_fixups = {
  .post_sfdp = en25qh256_post_sfdp_fixups,
};
```

It does look a lot cleaner :-)
I'll create the patch and do the rebase to the new kernel next.

Reply:
it IS tomorrow already - might as well do it now.. :grimacing: